### PR TITLE
copy - support recursive copying with remote_src

### PR DESCRIPTION
**Pull request by ptux**
##### SUMMARY

implement #14131

The followings will be supported.

* push remote src to remote dest with idempotence.
* supports check mode currently.
* supports symlink.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
copy.py

##### ANSIBLE VERSION

```
ansible 2.5 2.6 2.7
```

### CODE REVIEW

**Comment by ptux** [comment]
@abadger 
ready to review. 

@IPvSean 

ccing you since you are waiting this.

**Comment by samdoran** [comment]
Please create a changelog fragment. See [fragments](https://github.com/ansible/ansible/tree/stable-2.6/changelogs/fragments) for examples.

**Comment by abadger** [comment]
Looks like this has different behaviour from remote_src=False:

When I specify:
```
copy:
    src: /var/tmp/ansible-2.6.0
    dest: /var/tmp/ansible-copy
    remote_src: False
```
I end up with a directory /var/tmp/ansible-copy/ansible-2.6.0/ which has the contents of the src directory.

When I specify
```
copy:
    src: /var/tmp/ansible-2.6.0
    dest: /var/tmp/ansible-copy
    remote_src: True
```
I end up with /var/tmp/ansible-copy/ with the contents of the src directory.  The second time i run the command i end up with a second directory inside of there,  /var/tmp/ansible-copy/ansible-2.6.0/ , which also has the contents of the src directory.

This should be changed so that it matches the remote_src: false behaviour and the integration tests updated to account for it.

**Comment by ptux** [comment]
@samdoran 
thanks for noticing, again.

@abadger 
thanks for pointing out the inconsistency.
I'll fix to ensure consistency with `remote_src: false`

**Comment by ptux** [comment]
@abadger
ready to review!

**Comment by abadger** [comment]
Noticed a couple things.

* It looks like mode is not set correctly for directories.  Try with something like ```mode: u=rwX,go-rwx``` to see that files are assigned the correct mode but directories are not.
* It looks like follow might be affecting symlinks in an unintended manner.  (follow with copy is a bit complex to understand so that's understandable).  follow is only used for the following scenario:
  * /var/tmp/symlink is a symbolic link pointing to /var/tmp/otherfile
  * dest=/var/tmp/symlink
  * if follow=True, then we dereference /var/tmp/symlink and our copy will overwrite /var/tmp/otherfile
  * if follow=False, then our copy will overwrite /var/tmp/symlink
* We recently added local_follow for what I think you're code currently uses follow for.  Maybe you want to make use of local_follow and add an alias to it to be src_follow or something?

**Comment by ptux** [comment]
@abadger 

### about mode

I had suggested the following on  #41222.
Maybe you missed it.

> I was coding for handling mode, owner, group.
> I noticed it's NOT a good idea that supports mode for the remote recursive copy.
> Which I mean who wants to grant SAME permission to dir and file.

> So, I suggest that make remote recursive copy support owner and group only.
> Users cloud use file module to change permission for dir and file separately

So, I didn't implement mode for `remote_src: True`.
Also, I checked the behavior when `remote_src: False`.(I mean copy local dir recursively to remote )
It has the same problem you mentioned.
> files are assigned the correct mode but directories are not.

If you decide to **add** mode to `remote_src: True`, and **fix** mode in `remote_src: False`,
shall we open another issue and PR for these?

### about follow

I NOW comprehend the behavior of `follow`.
I'm getting confused about the goal...

Please let me think about that a little more.

**Comment by abadger** [comment]
yeah, if mode is currently not working for remote_src=false then we can fix both in a separate pr.  i thought i answered t your question about string directories and files in the same command but most not have submitted it.  the answer is that people can use symbolic modes to change permissions appropriately

```
mode: u=rwX,go=rX
```

**Comment by ptux** [comment]
@abadger 

I guess I'v misread the answer.
So, I'm going to open another issue(#44494) about `mode`, both remote_src=false  and remote_src=true.
And I'll implement it myself.

In this PR, I'll focus on `follow`.

**Comment by ptux** [comment]
@abadger

Ready to review!

- use local_follow to handler symbolic
- match the remote_src: false behavior

**Comment by ptux** [comment]
@abadger

This is just a friendly reminder.

**Comment by ptux** [comment]
@abadger 
This is a friendly reminder.

File: test/integration/targets/copy/tasks/tests.yml [hunk 1]
```diff
@@ -1488,3 +1488,594 @@
...
+    user:
+      name: ptux
+      uid: 14131
+      state: present
+
+  - name: execute - create a group for test
+    group:
+      name: ptux
+      gid: 14131
+      state: present
```
Comment (abadger) [inline test/integration/targets/copy/tasks/tests.yml]: @mattclay do we already have a standard user and group for testing setting file ownership  (nobody maybe?)

Comment (mattclay) [reply test/integration/targets/copy/tasks/tests.yml]: No, we don't. If the test needs additional user(s) it should create them in a `block` and use `always` to remove them. The test should also have the `needs/root` alias.

Comment (ptux) [reply test/integration/targets/copy/tasks/tests.yml]: I was thinking the test is executed in a container.
Everything will go when the test over. Am I wrong?

Also, What does [needs/root alias] mean?

**Comment by abadger** [comment]
Copy already has it so that's fine:
https://github.com/ansible/ansible/blob/devel/test/integration/targets/copy/aliases

File: test/integration/targets/copy/tasks/tests.yml (hunk 1 above)
Comment (mattclay) [reply test/integration/targets/copy/tasks/tests.yml]: Our CI system executes the tests in a container. However, we support running them in other environments and do not require the use of containers, so tests should clean up after themselves.

Comment (ptux) [reply test/integration/targets/copy/tasks/tests.yml]: got it.

**Comment by ptux** [comment]
> If the test needs additional user(s) it should create them in a block and use always to remove them. 

fixed.

File: test/integration/targets/copy/tasks/tests.yml (hunk 1 above)
Comment (mattclay) [reply test/integration/targets/copy/tasks/tests.yml]: Use a variable for the user and group and set them to a descriptive prefix like "ansible_copy_test_" and a random number. That will nearly eliminate any possibility of it matching an existing user, while making it more obvious where it came from if still present when the test exists (ex: test traceback).

**Comment by ptux** [comment]
> Use a variable for the user and group and set them to a descriptive prefix like "ansible_copy_test_" and a random number.

fixed.

File: changelogs/fragments/copy-recursive-remote-src.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+- copy module - implement that allow copy module to work with recursive and remote_src #14131
```
Comment (mattclay) [inline changelogs/fragments/copy-recursive-remote-src.yml]: copy - support recursive copying with remote_src

File: lib/ansible/modules/files/copy.py [hunk 3]
```diff
@@ -78,7 +78,7 @@
     description:
       - If C(no), it will search for I(src) at originating/master machine.
       - If C(yes) it will go to the remote/target machine for the I(src). Default is C(no).
-      - Currently I(remote_src) does not support recursive copying.
+      - I(remote_src) supports recursive copying in version 2.7 and higher.
```
Comment (mattclay) [inline lib/ansible/modules/files/copy.py]: - I(remote_src) supports recursive copying as of version 2.8.

File: changelogs/fragments/copy-recursive-remote-src.yml (hunk 2 above)
Comment (ptux) [reply changelogs/fragments/copy-recursive-remote-src.yml]: concise statement!

File: lib/ansible/modules/files/copy.py (hunk 3 above)
Comment (ptux) [reply lib/ansible/modules/files/copy.py]: concise statement!

**Comment by ptux** [comment]
@abadger 
@mattclay 
 
Is it good to merge?

**Comment by ptux** [comment]
ready_for_review

**Comment by ptux** [comment]
bot_status

**Comment by d4rkd0s** [comment]
Just ran into this issue today, very exciting to see this is nearing merge. Awesome work @ptux

**Comment by danihodovic** [comment]
What Ansible version would this land in? Would it be backported?

**Comment by ptux** [comment]
@abadger
@mattclay

I think you have been very busy because of the AnsibleFest.
This is just a friendly reminder.

**Comment by ptux** [comment]
Is anybody awake?

**Comment by abadger** [comment]
Thanks @ptux.  Yeah, ansiblefest and the 2.7.0 release took time away from other things.  This looks good.  merging.

**Comment by jonathanspw** [comment]
Is there a target current version for this to be added in or will we not see it until 2.8?

**Comment by sivel** [comment]
@neonardo1 This was merged after the release of 2.7.0, and as we do not backport new features, this will appear in the 2.8 release.

**Comment by IPvSean** [comment]
thanks @ptux !!

**Pull request merged**